File: packages/core/ios/WebViewHandler.swift
//
// Copyright 2025 Splunk Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import Foundation
import React
import WebKit
import SplunkAgent

/// Handles WebView integration with Browser RUM.
@objcMembers
class WebViewHandler: NSObject {

  private weak var bridge: RCTBridge?

  init(bridge: RCTBridge?) {
    self.bridge = bridge
    super.init()
  }

  func setBridge(_ bridge: RCTBridge?) {
    self.bridge = bridge
  }

  func integrateWithBrowserRum(
    _ viewTag: NSNumber,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let bridge = self.bridge else {
      reject("E_WEBVIEW", "Bridge not available", nil)
      return
    }

    DispatchQueue.main.async {
      guard let uiManager = bridge.module(forName: "UIManager") as? RCTUIManager else {
        reject("E_WEBVIEW", "UIManager not available", nil)
        return
      }

      uiManager.addUIBlock { [weak self] _, viewRegistry in
        guard let self = self else {
          reject("E_WEBVIEW", "Handler deallocated", nil)
          return
        }

        guard let viewRegistry = viewRegistry else {
          reject("E_WEBVIEW", "View registry not available", nil)
          return
        }

        guard let view = viewRegistry[viewTag] else {
          reject("E_WEBVIEW", "View not found for tag \(viewTag)", nil)
          return
        }

        if let webView = self.findWKWebView(in: view) {
          SplunkRum.shared.webViewNativeBridge.integrateWithBrowserRum(webView)
          resolve(nil)
        } else {
          reject("E_WEBVIEW", "View is not a WKWebView", nil)
        }
      }
    }
  }

  /// Recursively searches for a WKWebView in the view hierarchy.
  private func findWKWebView(in view: UIView) -> WKWebView? {
    if let webView = view as? WKWebView {
      return webView
    }
    for subview in view.subviews {
      if let found = findWKWebView(in: subview) {
        return found
      }
    }
    return nil
  }
}

File: packages/core/android/src/main/kotlin/com/splunk/otel/reactnative/WebViewHandler.kt
/*
 * Copyright 2025 Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.otel.reactnative

import android.webkit.WebView
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.fabric.FabricUIManager
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.UIManagerModule
import com.splunk.rum.integration.webview.WebViewNativeBridge

/**
 * Handles WebView integration.
 */
class WebViewHandler(private val reactContext: ReactApplicationContext) {

  fun integrateWithBrowserRum(viewTag: Double, promise: Promise) {
    try {
      UiThreadUtil.runOnUiThread {
        try {
          val tag = viewTag.toInt()
          val uiManager = UIManagerHelper.getUIManager(reactContext, tag)
          
          val view = when (uiManager) {
            is FabricUIManager -> uiManager.resolveView(tag)
            is UIManagerModule -> uiManager.resolveView(tag)
            else -> null
          }

          if (view is WebView) {
            WebViewNativeBridge.instance.integrateWithBrowserRum(view)
          }

          promise.resolve(null)
        } catch (e: Throwable) {
          promise.reject("E_WEBVIEW", e)
        }
      }
    } catch (t: Throwable) {
      promise.reject("E_WEBVIEW", t)
    }
  }
}
